crates/sui-core/src/transaction_orchestrator.rs

```diff
@@ -448,21 +459,29 @@ where
     ) -> Result<QuorumTransactionResponse, QuorumDriverError> {
         debug!("Using TransactionDriver for transaction {:?}", tx_digest);
         // Add transaction to WAL log for TransactionDriver path
-        if self
+        let is_new_transaction = self
             .pending_tx_log
             .write_pending_transaction_maybe(verified_transaction)
             .await
             .map_err(|e| {
                 warn!(?tx_digest, "QuorumDriverInternalError: {e:?}");
                 QuorumDriverError::QuorumDriverInternalError(e)
-            })?
-        {
+            })?;
+
+        if !is_new_transaction {
             debug!(
                 ?tx_digest,
-                "Added transaction to WAL log for TransactionDriver"
+                "Transaction already in pending_tx_log, returning TransactionPendingExecution"
             );
+            // Return the special error to signal that we should wait for effects
+            return Err(QuorumDriverError::TransactionPendingExecution);
```

Comment:
Isn't possible for the `submit_with_transaction_driver` feature to get cancelled (ex because the `WAIT_FOR_FINALITY_TIMEOUT_SECS` has been hit) and effectively leave the transaction in WAL? Then if the user retries to submit the same tx they'll see the `QuorumDriverError::TransactionPendingExecution` , although we have no guarantee on which state the tx is (maybe we never managed to successfully submit it? )

Reply:
Good catch!